### PR TITLE
Use plugin-pom 2.12, Make buildable with Java8

### PR DESCRIPTION
**Pull request by ikedam**
Current pom.xml have problems in building (and releasing) for accessing maven.jenkins-ci.org as it is no longer available.
This change makes rebuild-plugin use plugin-pom: https://github.com/jenkinsci/plugin-pom

* Good points:
    * Fixes [INFRA-588](https://issues.jenkins-ci.org/browse/INFRA-588)
        * Replace maven.jenkins-ci.org to repo.jenkins-ci.org.
        * Using plugin-pom makes it easy to catch up with those kinds of fixes in future.
* Bad points:
    * Some workarounds are required to target old Jenkins.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -17,15 +14,19 @@
     <packaging>hpi</packaging>
     <name>Rebuilder</name>
 
+    <properties>
+        <!-- Since rebuild plugin is using some functions available in
+        Jenkins version 1.405 ,version is updating from 1.400 to 1.405-->
+        <jenkins.version>1.481</jenkins.version>
+        <jenkins-test-harness.version>1.481</jenkins-test-harness.version>
+        <java.level>5</java.level>
```
Comment (ikedam) [inline pom.xml]: Jenkins < 1.520 targets Java5.
And rebuild plugin doesn't use features of Java6, anyway.

File: pom.xml [hunk 2]
```diff
@@ -17,15 +14,19 @@
...
+
     <repositories>
         <repository>
             <id>repo.jenkins-ci.org</id>
             <url>http://repo.jenkins-ci.org/public/</url>
         </repository>
-        <repository>
-            <id>maven.jenkins-ci.org</id>
-            <url>http://maven.jenkins-ci.org/content/repositories/releases</url>
-        </repository>
```
Comment (ikedam) [inline pom.xml]: No longer available.

File: pom.xml [hunk 3]
```diff
@@ -89,37 +90,54 @@
...
+                                       <!-- dependencies via jenkins-core-1.481 -->
+                                       <exclude>org.kohsuke:asm3</exclude>
+                                       <exclude>org.mindrot:jbcrypt</exclude>
+                                   </excludes>
+                               </enforceBytecodeVersion>
+                           </rules>
+                       </configuration>
+                   </execution>
+               </executions>
+           </plugin>
```
Comment (ikedam) [inline pom.xml]: Workarounds to build with older Jenkins core.

File: pom.xml [hunk 4]
```diff
@@ -89,37 +90,54 @@
...
-            <plugin>
-                <artifactId>maven-jar-plugin</artifactId>
-                <executions>
-                    <execution>
-                        <goals>
-                            <goal>test-jar</goal>
-                        </goals>
-                    </execution>
-                </executions>
-            </plugin>
```
Comment (ikedam) [inline pom.xml]: This is configured in plugin-pom.

File: pom.xml [hunk 5]
```diff
@@ -129,12 +147,6 @@
             </extension>
         </extensions>
     </build>
-    <distributionManagement>
-        <repository>
-            <id>maven.jenkins-ci.org</id>
-            <url>http://maven.jenkins-ci.org:8081/content/repositories/releases</url>
-        </repository>
-    </distributionManagement>
```
Comment (ikedam) [inline pom.xml]: This is configured in plugin-pom. (and it uses repo.jenkins-ci.org)

File: pom.xml [hunk 6]
```diff
@@ -154,30 +166,12 @@
...
-            <plugin>
-                <groupId>org.codehaus.mojo</groupId>
-                <artifactId>findbugs-maven-plugin</artifactId>
-                <version>2.0.1</version>
-                <configuration>
-                    <xmlOutput>true</xmlOutput>
-                    <!-- Optional derectory to put findbugs xdoc xml report -->
-                    <!--<xmlOutputDirectory>target/site</xmlOutputDirectory>-->
-                </configuration>
-            </plugin>
```
Comment (ikedam) [inline pom.xml]: This is configured in plugin-pom.

File: pom.xml [hunk 7]
```diff
@@ -154,30 +166,12 @@
...
         </plugins>
     </reporting>
-    <dependencies>
-        <dependency>
-            <groupId>com.google.guava</groupId>
-            <artifactId>guava</artifactId>
-            <version>12.0-rc1</version>
-            <scope>provided</scope>
-        </dependency>
-    </dependencies>
```
Comment (ikedam) [inline pom.xml]: com.google.guava:guava:jar:11.0.1 is provided via Jenkins core, and rebuild-plugin doesn't seem require features of 12.0-rc1.

File: src/main/resources/com/sonyericsson/rebuild/RebuildAction/BooleanParameterValue.jelly [hunk 8]
```diff
@@ -24,6 +24,7 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 -->
 
+<?jelly escape-by-default='true'?>
```
Comment (ikedam) [inline src/main/resources/com/sonyericsson/rebuild/RebuildAction/BooleanParameterValue.jelly]: Plugin-pom run tests to verify that escape-by-default='true' is set for all jelly files.

File: src/main/java/com/sonyericsson/rebuild/RebuildParameterProvider.java [hunk 9]
```diff
@@ -48,6 +48,7 @@
  *       other than your custom {@link ParameterValue}.
  *       There are two recommended ways to set values to {@link RebuildParameterPage}:
  *       <table>
+ *           <caption>Recommended values for {@link RebuildParameterPage}</caption>
```
Comment (ikedam) [inline src/main/java/com/sonyericsson/rebuild/RebuildParameterProvider.java]: Fix to build with Java8.

File: pom.xml [hunk 10]
```diff
@@ -5,10 +5,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <!-- Since rebuild plugin is using some functions available in
-        Jenkins version 1.405 ,version is updating from 1.400 to 1.405-->
-        <version>1.481</version>
-        <relativePath>../pom.xml</relativePath>
+        <version>2.12</version>
```
Comment (oleg-nenashev) [inline pom.xml]: Maybe 2.33 to pick the upperBounds enforcer?

File: pom.xml (hunk 1 above)
Comment (oleg-nenashev) [reply pom.xml]: I would probably bump it to something reasonably new like 1.609.3 (Java 6). It pulls in more annotations in the core && seems to be reasonable for testing purposes. According to the usage stats, it's pretty fine: http://stats.jenkins.io/pluginversions/rebuild.html

File: pom.xml [hunk 11]
```diff
@@ -17,15 +14,19 @@
...
+        Jenkins version 1.405 ,version is updating from 1.400 to 1.405-->
+        <jenkins.version>1.481</jenkins.version>
+        <jenkins-test-harness.version>1.481</jenkins-test-harness.version>
+        <java.level>5</java.level>
+    </properties>
+
     <repositories>
         <repository>
             <id>repo.jenkins-ci.org</id>
             <url>http://repo.jenkins-ci.org/public/</url>
```
Comment (oleg-nenashev) [inline pom.xml]: https

File: pom.xml [hunk 12]
```diff
@@ -73,8 +74,8 @@
 
     <pluginRepositories>
         <pluginRepository>
-            <id>maven.jenkins-ci.org</id>
-            <url>http://maven.jenkins-ci.org/content/repositories/releases</url>
+          <id>repo.jenkins-ci.org</id>
+          <url>http://repo.jenkins-ci.org/public/</url>
```
Comment (oleg-nenashev) [inline pom.xml]: https

File: pom.xml [hunk 13]
```diff
@@ -89,37 +90,54 @@
         </pluginManagement>
         <plugins>
             <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-compiler-plugin</artifactId>
+                <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <!-- jenkins-test-harness < 1.545 doesn't support concurrent tests. -->
```
Comment (oleg-nenashev) [inline pom.xml]: I would suggest bumping it to the version in Parent POM in order to pick more injected tests

**Comment by GLundh** [comment]
This PR have some merge conflicts. Please rebase or re-merge from master.

As Oleg suggests, also increase Jenkins version 1.609.3.

**Comment by ikedam** [comment]
Closed the request as it's completely outdated.
The latest plugin-pom is now 2.36 and provides upper bound checks for dependencies. And we also need Jenkinsfile to run tests on ci.jenkins.io.
It requires some more work and I don't think I can restart it soon.
You can create a brand-new pull request.

**Pull request closed without merging**